### ISSUE TITLE
Dynamic configuration breaks installation on standalone worker

### ISSUE THREAD

**Issue by ryantking** [k0sproject/k0s#2753]
### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
Linux 5.15.85-1-pve #1 SMP PVE 5.15.85-1 (2023-02-01T00:00Z) x86_64 GNU/Linux
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

### Version

v1.26.0+k0s.0

### Sysinfo

<details><summary>`k0s sysinfo`</summary>
```text
Machine ID: "fc70219997392e455981de2719c1c8aa08ed65329c510ff5f6411b5362bff479" (from machine) (pass)
Total memory: 188.6 GiB (pass)
Disk space available for /var/lib/k0s: 7.2 GiB (pass)
Operating system: Linux (pass)
  Linux kernel release: 5.15.85-1-pve (pass)
  Max. file descriptors per process: current: 524288 / max: 524288 (pass)
  Executable in path: modprobe: /usr/sbin/modprobe (pass)
  /proc file system: mounted (0x9fa0) (pass)
  Control Groups: version 2 (pass)
    cgroup controller "cpu": available (pass)
    cgroup controller "cpuacct": available (via cpu in version 2) (pass)
    cgroup controller "cpuset": available (pass)
    cgroup controller "memory": available (pass)
    cgroup controller "devices": available (assumed) (pass)
    cgroup controller "freezer": available (assumed) (pass)
    cgroup controller "pids": available (pass)
    cgroup controller "hugetlb": available (pass)
    cgroup controller "blkio": available (via io in version 2) (pass)
  CONFIG_CGROUPS: Control Group support: built-in (pass)
    CONFIG_CGROUP_FREEZER: Freezer cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_PIDS: PIDs cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_DEVICE: Device controller for cgroups: built-in (pass)
    CONFIG_CPUSETS: Cpuset support: built-in (pass)
    CONFIG_CGROUP_CPUACCT: Simple CPU accounting cgroup subsystem: built-in (pass)
    CONFIG_MEMCG: Memory Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_HUGETLB: HugeTLB Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_SCHED: Group CPU scheduler: built-in (pass)
      CONFIG_FAIR_GROUP_SCHED: Group scheduling for SCHED_OTHER: built-in (pass)
        CONFIG_CFS_BANDWIDTH: CPU bandwidth provisioning for FAIR_GROUP_SCHED: built-in (pass)
    CONFIG_BLK_CGROUP: Block IO controller: built-in (pass)
  CONFIG_NAMESPACES: Namespaces support: built-in (pass)
    CONFIG_UTS_NS: UTS namespace: built-in (pass)
    CONFIG_IPC_NS: IPC namespace: built-in (pass)
    CONFIG_PID_NS: PID namespace: built-in (pass)
    CONFIG_NET_NS: Network namespace: built-in (pass)
  CONFIG_NET: Networking support: built-in (pass)
    CONFIG_INET: TCP/IP networking: built-in (pass)
      CONFIG_IPV6: The IPv6 protocol: built-in (pass)
    CONFIG_NETFILTER: Network packet filtering framework (Netfilter): built-in (pass)
      CONFIG_NETFILTER_ADVANCED: Advanced netfilter configuration: built-in (pass)
      CONFIG_NETFILTER_XTABLES: Netfilter Xtables support: module (pass)
        CONFIG_NETFILTER_XT_TARGET_REDIRECT: REDIRECT target support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_COMMENT: "comment" match support: module (pass)
        CONFIG_NETFILTER_XT_MARK: nfmark target and match support: module (pass)
        CONFIG_NETFILTER_XT_SET: set target and match support: module (pass)
        CONFIG_NETFILTER_XT_TARGET_MASQUERADE: MASQUERADE target support: module (pass)
        CONFIG_NETFILTER_XT_NAT: "SNAT and DNAT" targets support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: "addrtype" address type match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_CONNTRACK: "conntrack" connection tracking match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_MULTIPORT: "multiport" Multiple port match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_RECENT: "recent" match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_STATISTIC: "statistic" match support: module (pass)
      CONFIG_NETFILTER_NETLINK: module (pass)
      CONFIG_NF_CONNTRACK: Netfilter connection tracking support: module (pass)
      CONFIG_NF_NAT: module (pass)
      CONFIG_IP_SET: IP set support: module (pass)
        CONFIG_IP_SET_HASH_IP: hash:ip set support: module (pass)
        CONFIG_IP_SET_HASH_NET: hash:net set support: module (pass)
      CONFIG_IP_VS: IP virtual server support: module (pass)
        CONFIG_IP_VS_NFCT: Netfilter connection tracking: built-in (pass)
      CONFIG_NF_CONNTRACK_IPV4: IPv4 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_REJECT_IPV4: IPv4 packet rejection: module (pass)
      CONFIG_NF_NAT_IPV4: IPv4 NAT: unknown (warning)
      CONFIG_IP_NF_IPTABLES: IP tables support: module (pass)
        CONFIG_IP_NF_FILTER: Packet filtering: module (pass)
          CONFIG_IP_NF_TARGET_REJECT: REJECT target support: module (pass)
        CONFIG_IP_NF_NAT: iptables NAT support: module (pass)
        CONFIG_IP_NF_MANGLE: Packet mangling: module (pass)
      CONFIG_NF_DEFRAG_IPV4: module (pass)
      CONFIG_NF_CONNTRACK_IPV6: IPv6 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_NAT_IPV6: IPv6 NAT: unknown (warning)
      CONFIG_IP6_NF_IPTABLES: IP6 tables support: module (pass)
        CONFIG_IP6_NF_FILTER: Packet filtering: module (pass)
        CONFIG_IP6_NF_MANGLE: Packet mangling: module (pass)
        CONFIG_IP6_NF_NAT: ip6tables NAT support: module (pass)
      CONFIG_NF_DEFRAG_IPV6: module (pass)
    CONFIG_BRIDGE: 802.1d Ethernet Bridging: built-in (pass)
      CONFIG_LLC: built-in (pass)
      CONFIG_STP: built-in (pass)
  CONFIG_EXT4_FS: The Extended 4 (ext4) filesystem: built-in (pass)
  CONFIG_PROC_FS: /proc file system support: built-in (pass)
```
</details>

### What happened?

I'm noticing two distinct issues with configuring k0s. Although I encountered the second one while working around the first one.

#### Main issue with dynamic configuration

The primary issue that I am facing is that I cannot install k0s using k0sctl on dedicated workers when `dynamicConfig` is enabled in my `k0sctl.yaml` file. I'm using a four node setup as follows:
1. controller
2. controller+worker
3. controller+worker
4. worker

The controller and worker services come up fine on nodes 1-3, but node 4 never comes online. Using the debug flag, I've found what looks to me like the main issue. The worker service is attempting to read a config map called `worker-config-default-1.26`, which I understand to be the old name for what is now `kubelet-config-default-1.26`. With dynamic configuration turned off, this error either does not occur or does not prevent the worker from joining the cluster.

#### Secondary issue with k0sctl config

Since dynamic configuration was not working, I've tried making changes such as enabling OpenEBS in my `k0sctl.yaml` file then re-rerunning `k0sctl apply`, but the output looks like the new config has been applied to each controller, but the changes are not reflected in cluster state.

#### Final workaround

I was able to get the dynamic configuration working by manually adding `--enable-dynamic-config` to the already installed controller services, reload systemd, then restart all the controller units. Dynamic configuration (both via `k0sctl` on my machine and `k0s` on a controller node) then worked, but did not have my initial `k0sctl` config so I had to put that back, specifically my `externalAddress` and `sans` values to get the HA control plane to work. Now the dynamic configuration seems to be working as expected. I still cannot make changes by re-applying my `k0sctl.yaml` file.

### Steps to reproduce

1. Create a new `k0sctl` config with at least one standalone worker and `dynamicConfig` set to true. To more closely emulate my setup also use ah HA control plane with one standalone controller and two nodes with both a controller and ` worker.` Optionally add more debug flags. The HA control plane requires the addition of the `externalAddress` and `sans` config options.
2. Run `k0sctl apply` with the configuration.
3. Watch the logs on the standalone worker, it should just hang indefinitely. With the debug flag it shows a permissions error about the missing `worker-config-default-1.26` error, which I assume is due to it not existing at all. I've seen that error sometimes without the debug flag, but not consistently.
4. Let `k0sctl apply` timeout or SIGINT it.
5. Reset the nodes through `k0sctl reset` or some other means.
6. Disable `dynamicConfig` in the `k0sctl` configuration file.
7. Run `k0sctl apply` again with the new configuration.
8. Observe it install correctly.
9. Make a change to the `k0sctl` configuration file that will have easily observable side effects. I enabled OpenEBS personally via `spec.extensions.storage`.
10. Re-run `k0sctl apply`.
11. Look for the side effect of the new configuration, it should not be there. For me, I saw that `kubectl get sc` did not return the local volume provisioner.
12. SSH into each control node and add `--enable-dynamic-config` to the end of the `ExecStart` line in `/etc/systemd/system/k0scontroller.service`. The run `systemctl daemon-reload` and `systemctl restart k0scontroller`.
13. Run `k0sctl config edit`. Note that the configuration is missing any manually set values from the initial/edited `k0sctl` config.
14. Edit the dynamic configuration and observe the changes take effect. Especially important if you are using an HA control plane as controllers will start crashing without the load balancer address.

### Expected behavior

1. Installing with `dynamicConfig: true` should work.
2. Running `k0sctl apply` with an update config file should apply the changes to the cluster.

### Actual behavior

1. Installing with `dynamicConfig: true` breaks standalone workers joining the cluster.
2. Running `k0sctl apply` with an update config file does not appear to change cluster state.

### Screenshots and logs

I hope I did a decent job describing the errors I saw, most of the logs were noise from konnectivity and such. I can dig up specific messages if that would help.

### Additional context

This cluster is running inside of LXCs in Proxmox VE. The controller-only node is in an unprivileged LXC while the worker nodes are in privileged LXCs with specific kubes workarounds (creating `kmsg`, disabling app armor, cgroups, pushing the kernel config, enabling nesting). I ran a similar setup with a non-HA control plane for a few weeks with no issues once I got the initial install working.

**Comment by twz123** [comment]
Thanks for reporting! The issue you are facing occurs when using an external load balancer (i.e. the `api.externalAddress` config setting) in conjunction with dynamic configuration. This is a bug in k0s and such a cluster will be broken, as you found out the hard way. #2787 fixes that.

> The worker service is attempting to read a config map called `worker-config-default-1.26`, which I understand to be the old name for what is now `kubelet-config-default-1.26`.

It's actually the other way round. `worker-config` is the new name, `kubelet-config` the old one.

> I was able to get the dynamic configuration working by manually adding `--enable-dynamic-config` to the already installed controller services, reload systemd, then restart all the controller units.

That probably only gives a somewhat working cluster. The API endpoint reconciliation (i.e. updating the contents of the default/kubernetes Endpoints resource) will probably still not work, but will remain at what the non-dynamically configured cluster has written into it, which is enough to join new workers.

> I hope I did a decent job describing the errors I saw, [...]

Yep! Thanks again for putting effort into collecting all of this :+1:

> 2. Running `k0sctl apply` with an update config file should apply the changes to the cluster.

When using dynamic configuration, you should be able to change the cluster configuration via `kubectl -n kube-system edit clusterconfig k0s`, no need for an additional `k0sctl` run. Whether or not `k0sctl` should update the configuration in this case is another story. Feel free to open up a discussion [at the forums](https://forums.k8slens.dev/) (or another GitHub issue) about that.

**Comment by kke** [comment]
> Whether or not k0sctl should update the configuration in this case is another story.

Yes it was given a lot of thought, but it would be very difficult to track which changes have been made outside of k0sctl, what to overwrite, what to keep, and so on. It would pretty much defeat the purpose of dynamic config anyway. If you want k0sctl to configure your cluster, it's better not to enable dynamic config.

K0sctl will warn the user that this is the last time k0sctl is going to touch your cluster config ever when dynamic config is enabled.

**Comment by ryantking** [comment]
Thanks @twz123 and @kke for your thorough answers. I must've missed the `k0sctl` warning about repeat initializations not changing cluster state.

Only outstanding question I have is if my "hack" to get the dynamic config working will cause problems down the road (re: what @twz123 said about the endpoint resource)? This is a production cluster in very early stages so if I have to reset and rebuild it, I'd rather bite the bullet now. Alternatively, will the backup and restore functionality that `k0s` provides be sufficient for rebuilding my cluster when #2787 ships if I can tolerate a little downtime while I reset/apply/restore?

Thanks again for the help and an overall excellent project. After answering those questions feel free to close this unless you want to leave it open for tracking purposes.

**Comment by twz123** [comment]
> Only outstanding question I have is if my "hack" to get the dynamic config working will cause problems down the road (re: what @twz123 said about the endpoint resource)? This is a production cluster in very early stages so if I have to reset and rebuild it, I'd rather bite the bullet now. Alternatively, will the backup and restore functionality that k0s provides be sufficient for rebuilding my cluster when https://github.com/k0sproject/k0s/pull/2787 ships if I can tolerate a little downtime while I reset/apply/restore?

IIUC this only affects only the default/kubernetes Endpoint resource. It won't be updated in any way. When using an external load balancer, the k0s controller performs a DNS lookup of the externalAddress and writes the results into default/kubernetes. This is the part that failed when dynamic config was enabled. So if your external address is not a DNS name, or your external address DNS name resolves to a stable IP, then you're more or less okay until the fix is shipped. The fix will be part of the v1.26.1+k0s.1 release, so whenever you upgrade your cluster to that version, the reconciliation should work again.

However, this is just an assessment of the problem itself. I didn't check too closely for any subsequent failures that might have been caused by it while I was fixing it. I'm pretty confident that there shouldn't be any, but I could be wrong,...
